runners/samza/src/main/java/org/apache/beam/runners/samza/metrics/SamzaMetricsContainer.java

```diff
@@ -81,15 +106,133 @@ public MetricsContainerStepMap getContainers() {
     return this.metricsContainers;
   }
 
-  /** Update Beam metrics to Samza metrics for the current step and global step. */
+  /**
+   * This is the public method for updating metrics. It either defers the update to the executor
+   * service or updates the metrics immediately based on the deferToExecutor configuration flag.
+   *
+   * @param stepName the step name for which metrics are being updated
+   */
   public void updateMetrics(String stepName) {
```

Comment:
Is this method `updateMetrics` the only source of overhead that is running  on the `main` thread?

Reply:
Not sure, but once we move this off the main thread perhaps we'll uncover more bottlenecks.